src/cc/BPF.cc

```diff
@@ -0,0 +1,418 @@
+
+namespace ebpf {
+
+std::string uint_to_hex(uint64_t value) {
+  std::stringstream ss;
+  ss << std::hex << value;
+  return ss.str();
+}
+
+std::string sanitize_str(
+  std::string str, bool (*validator)(char), char replacement = '_'
+) {
+  for (size_t i = 0; i < str.length(); i++)
+    if (!validator(str[i]))
+      str[i] = replacement;
+  return str;
+}
+
+BPF::BPF(
+  const std::string& bpf_program,
+  unsigned int debug_flag,
+  std::vector<std::string> cflags
+) :
+  bpf_module_(new BPFModule(debug_flag))
+{
+  const char* flags[cflags.size()];
+  for (size_t i = 0; i < cflags.size(); i++)
+    flags[i] = cflags[i].c_str();
+  if (bpf_module_->load_string(bpf_program, flags, cflags.size()) != 0)
+    throw std::runtime_error("Unable to initialize BPF program");
+};
+
+BPF::~BPF() {
+  for (auto it : kprobes_) {
+    auto res = detach_kprobe_event(it.first, it.second);
+    if (std::get<0>(res) != 0)
+      std::cerr << "Failed to detach kprobe event " << it.first
+                << " in BPF destructor: " << std::get<1>(res) << std::endl;
+  }
+
+  for (auto it : uprobes_) {
+    auto res = detach_uprobe_event(it.first, it.second);
+    if (std::get<0>(res) != 0)
+      std::cerr << "Failed to detach uprobe event " << it.first
+                << " in BPF destructor" << std::get<1>(res) << std::endl;
+  }
+
+  for (auto it : tracepoints_) {
+    auto res = detach_tracepoint_event(it.first, it.second);
+    if (std::get<0>(res) != 0)
+      std::cerr << "Failed to detach Tracepoint " << it.first
+                << " in BPF destructor" << std::get<1>(res) << std::endl;
+  }
+
+  for (auto it : perf_buffers_)
+    delete it.second;
+}
+
+StatusTuple BPF::attach_kprobe(
+  const std::string& kernel_func,
+  const std::string& probe_func,
+  bpf_attach_type attach_type,
+  pid_t pid, int cpu, int group_fd,
+  perf_reader_cb cb, void* cb_cookie
+) {
+  int probe_fd;
+  TRY2(load_func(probe_func, BPF_PROG_TYPE_KPROBE, probe_fd));
+
+  std::string probe_event = get_kprobe_event(kernel_func, attach_type);
+  if (kprobes_.find(probe_event) != kprobes_.end()) {
+    TRY2(unload_func(probe_func));
+    return mkstatus(-1, "kprobe %s already attached", probe_event.c_str());
+  }
+
+  std::string probe_event_desc = attach_type_prefix(attach_type);
+  probe_event_desc += ":kprobes/" + probe_event + " " + kernel_func;
+
+  void* res = bpf_attach_kprobe(
+    probe_fd,
+    probe_event.c_str(), probe_event_desc.c_str(),
+    pid, cpu, group_fd,
+    cb, cb_cookie
+  );
+
+  if (!res) {
+    TRY2(unload_func(probe_func));
+    return mkstatus(
+      -1, "Unable to attach %skprobe for %s using %s",
+      attach_type_debug(attach_type).c_str(),
+      kernel_func.c_str(), probe_func.c_str()
+    );
+  }
+
+  open_probe_t p = {};
+  p.reader_ptr = res;
+  p.func = probe_func;
+  kprobes_[probe_event] = std::move(p);
+  return mkstatus(0);
+}
+
+StatusTuple BPF::attach_uprobe(
+  const std::string& binary_path, const std::string& symbol,
```

Comment:
please reformat the whole thing similar to bpf_module.*
bcc doesn't have 80 char limit. Use the following style:
```
int foo(int arg1, int arg2,
        int arg3) {
}
```
which is pretty much google coding style.